core/src/main/java/org/apache/iceberg/FastAppend.java

```diff
@@ -81,15 +83,21 @@ public FastAppend appendFile(DataFile file) {
 
   @Override
   public FastAppend appendManifest(ManifestFile manifest) {
-    // the manifest must be rewritten with this update's snapshot ID
-    try (ManifestReader reader = ManifestReader.read(
-        ops.io().newInputFile(manifest.path()), ops.current().specsById())) {
-      OutputFile newManifestPath = manifestPath(manifestCount.getAndIncrement());
-      appendManifests.add(ManifestWriter.copyAppendManifest(reader, newManifestPath, snapshotId(), summaryBuilder));
+    Preconditions.checkArgument(!manifest.hasExistingFiles(), "Cannot append manifest with existing files");
+    Preconditions.checkArgument(!manifest.hasDeletedFiles(), "Cannot append manifest with deleted files");
+    Preconditions.checkArgument(manifest.snapshotId() == null, "Snapshot id must be assigned during commit");
+
+    // TODO: avoid reading manifests to simply get stats
+    try (ManifestReader reader = ManifestReader.read(manifest, ops.io(), ops.current().specsById())) {
```

Comment:
I think we need to collect more metadata while writing manifests so that we don't have to read manifests to simply get stats. Clearly, this kills all the benefits of inhering the snapshot id.

Reply:
Yes, we can keep the summary in `ManifestFile` instead.

Reply:
Actually, we can use the manifest's summary stats for top-level properties. Partition-level properties are optional so we should just not include them.

Reply:
Ideally, we would produce all of the summary stats, but the most important ones are `total-data-files`, `total-records`, and the `added-` or `deleted-` properties that are used to produce totals. I think it's okay to not write the `changed-partition-count` metrics if they require scanning the appended manifest.

I think my response was confusing because we keep additional summary information about each partition in our version. I can move that upstream if everyone wants it, but it can make the metadata files quite large. Without a use case for doing this upstream, I didn't think it was a good idea to make everyone's metadata significantly larger.

Reply:
Do you have a few examples of what stats you collect per partition? One of our customers was asking about some stats about what partitions were modified, for example.